File: library/src/main/java/com/snail/fileselector/Item.java
package com.snail.fileselector;

import androidx.annotation.NonNull;

import java.io.File;

/**
 * Created by zeng on 2017/3/1.
 */

class Item implements Comparable<Item> {
    String desc;
	File file;
	boolean checked;

	Item(File file, boolean checked) {
		this.file = file;
		this.checked = checked;
	}
	
	Item(String desc, File file, boolean checked) {
	    this(file, checked);
	    this.desc = desc;
    }

	@Override
	public int compareTo(@NonNull Item o) {
		if (file == null) {
		    return -1;
		} else if (o.file == null) {
		    return 1;
		}
		String s = CharacterParser.getSelling(file.getName());
		String s1 = CharacterParser.getSelling(o.file.getName());
		return s.compareTo(s1);
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (!(o instanceof Item)) return false;
		Item item = (Item) o;
		return file != null ? file.equals(item.file) : item.file == null;

	}

	@Override
	public int hashCode() {
		return file != null ? file.hashCode() : 0;
	}
}

File: library/src/main/java/com/snail/fileselector/Item.kt
package com.snail.fileselector

import java.io.File

/**
 * Created by zeng on 2017/3/1.
 */

internal class Item @JvmOverloads constructor(var file: File?, var checked: Boolean, var desc: String = "") : Comparable<Item> {

    override fun compareTo(other: Item): Int {
        if (file == null) {
            return -1
        } else if (other.file == null) {
            return 1
        }
        val s = CharacterParser.getSelling(file!!.name)
        val s1 = CharacterParser.getSelling(other.file!!.name)
        return s.compareTo(s1)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Item) return false
        val item = other as Item?
        return if (file != null) file == item!!.file else item!!.file == null

    }

    override fun hashCode(): Int {
        return if (file != null) file!!.hashCode() else 0
    }
}
